### PR TITLE
Handle empty tensors in cat and slice_assign ops

### PR DESCRIPTION
**Pull request by antimora**
Updated tensor API to treat empty tensors as no-ops in slice_assign and to filter out tensors with size 0 along the concatenation dimension in cat. Added tests to verify correct behavior when handling empty tensors and empty slice assignments. This prevents backend errors and ensures consistent results for edge cases.

### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

- Fixes https://github.com/tracel-ai/burn/issues/4090
- https://github.com/tracel-ai/burn/pull/4083
- https://github.com/tracel-ai/burn/issues/4052
- https://github.com/tracel-ai/burn/issues/4093

### Changes

- Empty slice assignments (where start == end or any dimension has size 0) now return the original tensor unchanged as a no-op
- The `cat` operation now filters out tensors with size 0 along the concatenation dimension before calling the backend
- Removed validation that prevented empty slice assignments, with handling moved to the API level

### Testing

Added new test cases for the change

### CODE REVIEW

File: crates/burn-tensor/src/tensor/api/base.rs [hunk 1]
```diff
@@ -1884,8 +1895,29 @@ where
...
+        // Empty tensors don't contribute to the output and would cause issues
+        // in backend implementations (e.g., division by zero in slice_assign).
+        let first_tensor = tensors.first().expect("Tensors should not be empty");
+        let device = first_tensor.device();
+        let mut shape = first_tensor.shape();
+
+        let non_empty_tensors: Vec<_> = tensors
+            .into_iter()
+            .filter(|t| t.shape().dims[dim] > 0)
+            .collect();
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/api/base.rs]: While we're iterating over the tensors to filter them, might as well map them to the primitives.

File: crates/burn-tensor/src/tensor/api/base.rs [hunk 2]
```diff
@@ -1884,8 +1895,29 @@ where
...
+            shape.dims[dim] = 0;
+            return Self::empty(shape, &device);
+        }
+
         Self::new(K::cat(
-            tensors.into_iter().map(|vector| vector.primitive).collect(),
+            non_empty_tensors
+                .into_iter()
+                .map(|vector| vector.primitive)
+                .collect(),
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/api/base.rs]: Which will remove this entirely
